### DISCUSSION TITLE
Network Circular layouts in Visus_On_Network

### DISCUSSION THREAD

**Discussion by jenny61599** [mesa/mesa#1769] [Q&A]
I am working with a modified Virus Model from the examples folder, and I was wondering if it is possible to visualize the network with a circular network in server.py? Below is my current code for generating graph G.

```
self.num_nodes = nodes
prob = avg_node_degree / self.num_nodes
self.G = nx.erdos_renyi_graph(n = self.num_nodes, p = prob) # generate random network

## This is my attempt in adding the circular layout in the network itself, but it's generating an error
#pos = nx.circular_layout(self.G)
# nx.set_node_attributes(0, pos, "pos")
        
 self.schedule = RandomActivation(self)

# add all the possible node values to G before it converts to a mesa network object
self.G.add_nodes_from(range(self.num_nodes))
self.schedule = mesa.time.RandomActivation(self)
```

As for the changes in server.py, I am still using mesa.visualization.NetworkModule() and mesa.visualization.ChartModule as in the original. I assume that to do a circular layout, it should be done in server.py instead of model.py?

**Comment by rht** [comment]
It is much harder to customize `mesa.visualization.NetworkModule()`. The new pure Python viz is much easier, however.
Here is an example for virus on network, https://github.com/projectmesa/mesa-examples/pull/56.

For your use case, I'd modify the app.py to have this
```python
def space_drawer(viz):
    space_fig = Figure()
    space_ax = space_fig.subplots()
    graph = viz.model.grid.G
    pos = nx.circular_layout(graph, seed=0)
    nx.draw(
        graph,
        ax=space_ax,
        pos=pos,
        **agent_portrayal(graph),
    )
    space_ax.set_axis_off()
    solara.FigureMatplotlib(space_fig, dependencies=[viz.model, viz.df])
```

and then modify the `JupyterViz` call to do this instead
```python
page = JupyterViz(
    VirusOnNetwork,
    model_params,
    measures=[
        make_plot,
        make_text(get_resistant_susceptible_ratio),
    ],
    name="Virus Model",
    space_drawer=space_drawer,
)
page  # noqa
```

This is a very new viz frontend, so I haven't made it even easier yet. But even for now, it is still easier than the default one.

**Reply by rht** [reply]
Forgot to add: you need to additionally import
```python
import solara
from matplotlib.figure import Figure
```

**Reply by rht** [reply]
To run the app.py, you do `solara run app.py`.

**Reply by rht** [reply]
https://github.com/projectmesa/mesa/discussions/1770 contains discussion on how to simplify custom space drawing.
